libraries/ms-drivers/src/ads1259_adc.c

```diff
@@ -121,3 +122,172 @@ StatusCode ads1259_get_conversion_data(Ads1259Storage *storage) {
+  prv_send_command(storage, ADS1259_STOP_READ_DATA_CONTINUOUS);
+  uint8_t payload[NUM_REGISTER_WRITE_COMM] = { (ADS1259_WRITE_REGISTER | ADS1259_ADDRESS_CONFIG0),
+                                               NUM_CONFIG_REGISTERS - 1, register_lookup[0],
+                                               register_lookup[1], register_lookup[2] };
+  spi_exchange(storage->spi_port, payload, NUM_REGISTER_WRITE_COMM, NULL, 0);
+  // sanity check that data was written correctly
+  uint8_t check_regs[] = { (ADS1259_READ_REGISTER | ADS1259_ADDRESS_CONFIG0),
+                           NUM_ADS1259_REGISTERS - 1 };
+  uint8_t reg_readback[NUM_ADS1259_REGISTERS];
+  spi_exchange(storage->spi_port, check_regs, sizeof(check_regs), reg_readback,
+               NUM_ADS1259_REGISTERS);
+  LOG_DEBUG("ads1259 post init register values:\n");
+  for (int i = 0; i < NUM_ADS1259_REGISTERS; i++) {
+    LOG_DEBUG("reg %d: 0x%x\n", i, reg_readback[i]);
+  }
+  return STATUS_CODE_OK;
+}
+
+// calculate check-sum based on page 29 of datasheet
+static Ads1259StatusCode prv_checksum(Ads1259Storage *storage) {
+  uint8_t sum = (uint8_t)(s_store.rx_data_lsb + s_store.rx_data_mid + s_store.rx_data_msb +
+                          ADS1259_CHECKSUM_OFFSET);
+  LOG_DEBUG("STORAGE CHK_SUM: %x\n", (uint8_t)s_store.chk_sum);
+  LOG_DEBUG("CALCULATED SUM %x\n", sum);
+  if ((uint8_t)s_store.chk_sum & CHK_SUM_FLAG_BIT) {
+    return ADS1259_STATUS_CODE_OUT_OF_RANGE;
+  }
+  if ((sum & ~(CHK_SUM_FLAG_BIT)) != ((uint8_t)s_store.chk_sum & ~(CHK_SUM_FLAG_BIT))) {
+    return ADS1259_STATUS_CODE_CHECKSUM_FAULT;
+  }
+  return ADS1259_STATUS_CODE_OK;
+}
+
+// using the amount of noise free bits based on the SPS and VREF calculate analog voltage value
+// 0x000000-0x7FFFFF positive range, 0xFFFFFF - 0x800000 neg range, rightmost is greatest magnitude
+static void prv_convert_data(Ads1259Storage *storage) {
+  double resolution = pow(2, s_num_usable_bits[ADS1259_DATA_RATE_SPS] - 1);
+  if (storage->conv_data.raw & RX_NEG_VOLTAGE_BIT) {
+    storage->reading = 0 - ((RX_MAX_VALUE - storage->conv_data.raw) >>
+                            (24 - s_num_usable_bits[ADS1259_DATA_RATE_SPS])) *
+                               EXTERNAL_VREF_V / resolution;
+  } else {
+    storage->reading = (storage->conv_data.raw >> (24 - s_num_usable_bits[ADS1259_DATA_RATE_SPS])) *
+                       EXTERNAL_VREF_V / (resolution - 1);
+  }
+}
+
+static void prv_conversion_callback(SoftTimerId timer_id, void *context) {
+  Ads1259Storage *storage = (Ads1259Storage *)context;
+  Ads1259StatusCode code;
+  uint8_t payload[] = { ADS1259_READ_DATA_BY_OPCODE };
+  spi_exchange(storage->spi_port, payload, 1, (uint8_t *)&storage->rx_data, NUM_ADS_RX_BYTES);
+  code = prv_checksum(storage);
+  (*storage->handler)(code, storage->error_context);
+
+  storage->conv_data.MSB = (uint8_t)s_store.rx_data_msb;
+  storage->conv_data.MID = (uint8_t)s_store.rx_data_mid;
+  storage->conv_data.LSB = (uint8_t)s_store.rx_data_lsb;
```

Comment:
could we actually change the structure of this to just insert the final converted value from python into `storage->reading`? my logic is that anytime we want to mock with this we'll want to insert a real voltage value, but it's much harder to try to insert a bunch of bytes (since we'd basically have to do the conversion backwards). So basically the last line of `prv_convert_data()` would be `storage->reading = s_store->reading;`, but also lemme know your thoughts on this

Reply:
We could, but wouldn't that mess with the checksum? Since it's calculated directly from `storage->rx_data` which has the `msb`, `lsb`, and `mid` elements.

Reply:
I don't think so, the checksum is internal to the driver only, so if you return the inserted reading at the end of `prv_convert_data` it shouldn't affect anything other than external driver users

Reply:
Ah yeah my bad I interpreted that as replacing the 3 stores as one `reading` store.